### PR TITLE
Adds the support for the GDDP reader

### LINKED ISSUE #989: Add the support for downscaled NASA Earth Exchange (NEX) Global Daily Downscaled Projections (GDDP) data as forcing
The NASA Earth Exchange (NEX) Global Daily Downscaled Projections (GDDP) dataset, or NEX-GDDP-CMIP6, was created to assist the science community in conducting studies of climate change impacts at local to regional scales and to enhance public understanding of possible future climate patterns at the spatial scale of individual towns, cities, and watersheds.

The NEX-GDDP-CMIP6 dataset is comprised of global downscaled climate scenarios derived from the General Circulation Model (GCM) runs conducted under the Coupled Model Intercomparison Project Phase 6 (CMIP6) and across two of the four “Tier 1” greenhouse gas emissions scenarios known as Shared Socioeconomic Pathways (SSPs).

https://www.nccs.nasa.gov/services/data-collections/land-based-products/nex-gddp-cmip6

### PR DESCRIPTION
**Pull request by sujayvkumar**
Adds the support for the NEX GDDP climate data as forcing. The plugin
temporally disaggregates the daily GDDP data to subdaily resolution.

The code also updates includes updates to the parameter reading routines
so that in a multiprocessor mode, each process directly reads the subsetted
data slice. The older version of the code read the global data and subsetted it.

Resolves #989

### Description

Replace this text with a concise description of *what* was changed and *why*.

### CODE REVIEW

**Comment by sujayvkumar** [comment]
A testcase is at : /discover/nobackup/projects/lis/svkumar/CASI2/GSFC_run

**Comment by bmcandr** [comment]
Hi @sujayvkumar,

The logs in my test run are filled with warnings about the SW radiation exceeding a max value, but these messages are not in your logs. Example:

```
[INFO] reading..
 ./input/MET_FORCING/NEX-GDDP-CMIP6//TaiESM1/historical/r1i1p1f1//tas/tas_day_Ta
 iESM1_historical_r1i1p1f1_gn_1957.nc
[INFO] LIS cycle time: 06/01/1957 02:00:00
[INFO] LIS cycle time: 06/01/1957 03:00:00
[INFO] LIS cycle time: 06/01/1957 04:00:00
[INFO] LIS cycle time: 06/01/1957 05:00:00
[INFO] LIS cycle time: 06/01/1957 06:00:00
[INFO] LIS cycle time: 06/01/1957 07:00:00
[INFO] LIS cycle time: 06/01/1957 08:00:00
[INFO] LIS cycle time: 06/01/1957 09:00:00
[INFO] LIS cycle time: 06/01/1957 10:00:00
 [WARN] sw radiation too high!!
 [WARN] it is           0           3   1384.366
 [WARN] data1=   1384.366
 [WARN] data2=   1384.366
 [WARN] wts=  0.0000000E+00   1.000000
 [WARN] zw1=  0.0000000E+00 zw2=   1.000000
 [WARN] swt1=  0.0000000E+00 swt2=   1.000000
 [WARN] sw radiation too high!!
 [WARN] it is           0           8   1452.208
 [WARN] data1=   1452.208
 [WARN] data2=   1452.208
 [WARN] wts=  0.0000000E+00   1.000000
 [WARN] zw1=  0.0000000E+00 zw2=   1.000000
 [WARN] swt1=  0.0000000E+00 swt2=   1.000000
...
```

These warnings continue for ~6000 lines before LIS cycles the next timestep. Aside from these messages, our logs and restart outputs are identical. Any idea what is causing this?

**Comment by sujayvkumar** [comment]
@bmcandr  - I think I turned off these warning statements in my code during testing. You can ignore them and focus on the outputs.

File: lis/core/LIS_fileIOMod.F90 [hunk 1]
```diff
@@ -1,9 +1,7 @@
 !-----------------------BEGIN NOTICE -- DO NOT EDIT-----------------------
-! NASA Goddard Space Flight Center
-! Land Information System Framework (LISF)
-! Version 7.3
+! NASA Goddard Space Flight Center Land Information System (LIS) v7.2
 !
-! Copyright (c) 2020 United States Government as represented by the
+! Copyright (c) 2015 United States Government as represented by the
 ! Administrator of the National Aeronautics and Space Administration.
 ! All Rights Reserved.
```
Comment (bmcandr) [inline lis/core/LIS_fileIOMod.F90:6]: Update copyright notice

File: lis/core/LIS_fileIOMod.F90 [hunk 2]
```diff
@@ -2851,6 +2849,7 @@ end subroutine read2Ddata
 ! !INTERFACE: 
  subroutine LIS_readDomainConfigSpecs(segment_name, domain_info)
 ! !USES: 
+   use ESMF
```
Comment (bmcandr) [inline lis/core/LIS_fileIOMod.F90:2852]: Remove duplicate `use ESMF` to appease Sonarqube scans (see 9100a91)

File: lis/core/LIS_fileIOMod.F90 [hunk 3]
```diff
@@ -3203,17 +3202,20 @@ subroutine readparam_real_2d(n,pname,array)
...
-     array(:,:) = &
-          param(LIS_ews_halo_ind(n,LIS_localPet+1):&         
-          LIS_ewe_halo_ind(n,LIS_localPet+1), &
-          LIS_nss_halo_ind(n,LIS_localPet+1): &
-          LIS_nse_halo_ind(n,LIS_localPet+1))
+!     array(:,:) = &
+!          param(LIS_ews_halo_ind(n,LIS_localPet+1):&         
+!          LIS_ewe_halo_ind(n,LIS_localPet+1), &
+!          LIS_nss_halo_ind(n,LIS_localPet+1): &
+!          LIS_nse_halo_ind(n,LIS_localPet+1))
```
Comment (bmcandr) [inline lis/core/LIS_fileIOMod.F90:3218]: Remove commented out code like this (multiple files)

File: lis/metforcing/gddp/get_gddp.F90 [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+             ref_hourlyclimofile,           &
+             ferror)
+        
+        if(ferror.ge.1) then
+           gddp_struc(n)%gddptime2=time2
+        endif
+        call LIS_tick(dtime2,doy2,gmt2,yr2,mo2,da2,hr2,mn2,ss2,ts2)
+        if(try.gt.11)then
+           write(*,*)'error: GDDP data gap exceeds 10 days on file 2'
+           stop
```
Comment (bmcandr) [inline lis/metforcing/gddp/get_gddp.F90:245]: Write to `LIS_logunit` and call `LIS_endrun()`

File: lis/metforcing/gddp/get_gddp.F90 [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+             names,                         &
+             ref_dailyclimofile,            &
+             ref_hourlyclimofile,           &
+             ferror)
+
+        if(ferror.ge.1) gddp_struc(n)%gddptime1=time1
+        call LIS_tick(dtime1,doy1,gmt1,yr1,mo1,da1,hr1,mn1,ss1,ts1)
+        if(try.gt.11)then
+           write(*,*)'error: GDDP data gap exceeds 10 days on file 1'
+           stop
```
Comment (bmcandr) [inline lis/metforcing/gddp/get_gddp.F90:208]: Log to `LIS_logunit` and call `LIS_endrun()`

File: lis/metforcing/gddp/read_gddp.F90 [hunk 6]
```diff
@@ -0,0 +1,935 @@
...
+        input_data_1d(c1+(r-1)*input_nc) = input_data(c,r)
+        if(input_data(c,r).ne.LIS_rc%udef) then
+           input_bitmap(c1+(r-1)*input_nc) = .true.
+        endif
+     enddo
+  enddo
+  
+!  open(100,file='test_inp.bin',form='unformatted')
+!  write(100) input_data_1d
+!  close(100)
```
Comment (bmcandr) [inline lis/metforcing/gddp/read_gddp.F90:823]: Remove debugging code

File: lis/metforcing/gddp/read_gddp.F90 [hunk 7]
```diff
@@ -0,0 +1,935 @@
...
+        output_2d(i,j) = output_data(i+(j-1)*nc)
+     enddo
+  enddo
+
+!  print*, 'here'
+!  open(100,file='test.bin',form='unformatted')
+!  write(100) output_data
+!  write(100) output_2d
+!  close(100)
+!  stop
```
Comment (bmcandr) [inline lis/metforcing/gddp/read_gddp.F90:890]: Remove debugging code

File: lis/metforcing/gddp/read_gddp.F90 [hunk 8]
```diff
@@ -0,0 +1,935 @@
...
+  do r=1,nr
+     do c=1,nc
+        found = .true. 
+        if(LIS_domain(n)%gindex(c,r).ne.-1.and.output_2d(c,r).eq.-9999.0) then
+           rad = 1
+           found = .false.
+
+!           if(LIS_localPet.eq.996.and.c.eq.88.and.r.eq.13) then
+!              print*, 'mismatch found ',LIS_localPet, vid, rad, c,r, output_2d(c,r)
+!           endif
```
Comment (bmcandr) [inline lis/metforcing/gddp/read_gddp.F90:901]: Remove debugging code

File: lis/metforcing/gddp/read_gddp.F90 [hunk 9]
```diff
@@ -0,0 +1,935 @@
...
+                       found = .true.
+                       
+                       exit
+                    endif
+                 enddo
+                 if(found) exit
+              enddo
+!              if(LIS_localPet.eq.996.and.c.eq.88.and.r.eq.13) then
+!                 print*, 'up ',LIS_localPet, rad,output_2d(c,r),found
+!              endif
```
Comment (bmcandr) [inline lis/metforcing/gddp/read_gddp.F90:921]: Remove debugging code

File: lis/metforcing/gddp/read_gddp.F90 [hunk 10]
```diff
@@ -0,0 +1,935 @@
...
+              if(.not.found) rad = rad +1
+              if(rad.gt.maxrad) exit
+           enddo
+           
+        endif
+        if(.not.found) then
+           print*, 'unable to fill', LIS_localPet, vid,  c,r,output_2d(c,r),found
```
Comment (bmcandr) [inline lis/metforcing/gddp/read_gddp.F90:928]: Need more descriptive log message here (and write to `LIS_logunit`)

File: lis/metforcing/gddp/read_gddp.F90 [hunk 11]
```diff
@@ -0,0 +1,935 @@
...
+                                found = .true.
+                                exit
+                             endif
+                          enddo
+                          if(found) exit
+                       enddo
+                       if(.not.found) rad = rad+1
+                    enddo
+                 elseif(huss_inp(c,r).lt.0) then
+                    print*, 'negative huss ',huss_inp(c,r)
```
Comment (bmcandr) [inline lis/metforcing/gddp/read_gddp.F90:459]: Need more descriptive log message here (and write to `LIS_logunit`)

**Review by bmcandr — CHANGES_REQUESTED** [review]
Hi @sujayvkumar,

Our outputs are identical, but I identified several minor changes that should be made before merging (see review comments on PR). I'm happy to make these changes if you approve.

I will also add this reader to the documentation.

**Comment by bmcandr** [comment]
Hi @sujayvkumar,

I pushed the changes I recommended in my review. While updating the user guide, I noticed that the GDDP forcing scenario field is not validated after it is read from the lis.config file. Is `historical` the only option that is currently supported?

**Comment by bmcandr** [comment]
I went ahead and added a check for the supported forcing scenario and updated the lis.config documentation.

I reproduced the target outputs with the Intel 2021 compiler. Merging

**Review by bmcandr — APPROVED** [review]

**Comment by sujayvkumar** [comment]
@bmcandr  - I found some bugs and I have some updates to commit. Can you undo the merge and reopen the pull requests?

**Comment by bmcandr** [comment]
@sujayvkumar sorry about that. I just reverted the merge. Pull from your remote branch to get the changes I made, make your fixes, and then you can reopen the PR when it's ready.

**Pull request merged**